src/storage/aol/segment.rs

```diff
@@ -0,0 +1,532 @@
+            let part = &rec[..l];
+            let buf = &mut p.buf[p.alloc..]; // Create a mutable slice starting from p.alloc
+
+            // Encode the content of 'part' into 'buf'
+            encode_record(buf, rec.len(), part, i);
+
+            p.alloc += part.len() + RECORD_HEADER_SIZE; // Update the number of bytes allocated in the block
+            if p.is_full() {
+                self.flush_block(true)?;
+            }
+
+            rec = &rec[l..]; // Update the remaining bytes to be written
+
+            n += l;
+
+            i += 1;
+        }
+
+        // Write the remaining data to the block
+        if self.block.alloc > 0 {
+            self.flush_block(false)?;
+        }
+
+        Ok((offset, n))
+    }
+
+    /// Reads data from the segment at the specified offset.
+    ///
+    /// This method reads data from the segment starting from the given offset. It reads
+    /// from the underlying file if the offset is beyond the current block's buffer. The
+    /// read data is then copied into the provided byte slice `bs`.
+    ///
+    /// # Parameters
+    ///
+    /// - `bs`: A byte slice to store the read data.
+    /// - `off`: The offset from which to start reading.
+    ///
+    /// # Returns
+    ///
+    /// Returns the number of bytes read and any encountered error.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the provided offset is negative or if there is an I/O error
+    /// during reading.
+    pub(crate) fn read_at(&self, bs: &mut [u8], off: u64) -> io::Result<usize> {
+        if off > self.offset() {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Offset beyond current position",
+            ));
+        }
+
+        // Calculate buffer offset
+        let mut boff = 0;
+
+        let mut n = 0;
+        if off < self.file_offset {
+            // Read from the file
+            n = self.file.read_at(bs, self.file_header_offset + off)?;
+        } else {
+            boff = (off - self.file_offset) as usize;
+        }
+
+        let pending = bs.len() - n;
+        if pending > 0 {
+            let available = self.block.unwritten() - boff;
+            let read_chunk_size = std::cmp::min(pending, available);
+
+            if read_chunk_size > 0 {
+                let buf = &self.block.buf
+                    [self.block.alloc + boff..self.block.alloc + boff + read_chunk_size];
+                merge_slices(bs, buf);
+            }
+
+            if read_chunk_size == pending {
+                return Ok(n);
+            } else {
+                return Err(io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "Incomplete read",
+                ));
+            }
+        }
+
+        Ok(n)
+    }
+}
+
+impl Drop for Segment {
+    /// Attempt to fsync data on drop, in case we're running without sync.
+    fn drop(&mut self) {
+        self.sync().ok();
```

Comment:
May we want to log that? So we have a chance to know that at some point in the code, `sync` has not been called before the segment has been dropped?

Reply:
Ack, I am adding that